spring-xd-messagebus-spi/src/main/java/org/springframework/xd/dirt/integration/bus/MessageBusSupport.java

```diff
@@ -898,7 +898,11 @@ private void revertDirectBindingIfNecessary(Binding binding) {
 
 		@Override
 		public int selectPartition(Object key, int partitionCount) {
-			return Math.abs(key.hashCode()) % partitionCount;
+			int hashCode = key.hashCode();
+			if (hashCode == Integer.MIN_VALUE) {
+				hashCode = 0;
+			}
+			return Math.abs(hashCode) % partitionCount;
```

Comment:
Actually, I think the modulo should not be part of the strategy as it is always applied.

Even, on closer inspection, I think there is an issue with 
```
if (partition >= meta.partitionCount) {
			partition = partition % meta.partitionCount;
		}
		if (partition < 0) {
			partition = Math.abs(partition);
		}
```

=> What if partition is less than `-meta.partitionCount` ? The modulo should be always applied, then abs() IMHO